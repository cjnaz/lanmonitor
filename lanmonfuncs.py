#!/usr/bin/env python3
"""LAN monitor support functions
"""

# __version__ = "V1.0a 210515"

#==========================================================
#
#  Chris Nelson, 2021
#
# V1.1  210523  cmd timeout tweaks
# V1.0  210507  V1.0
# V1.0a 210515  Set timeouts to 1s for ping and 5s for ssh commands on remotes
#
# Changes pending
#   
#==========================================================

import sys
import os.path
import subprocess
import datetime
import time
import re
sys.path.append(os.path.join(os.path.dirname(os.path.realpath(__file__)), '../funcs3/'))
from funcs3 import logging, getcfg, snd_email, snd_notif

# Configs / Constants
NOTIF_SUBJ = "LAN Monitor"
RTN_PASS     = 0
RTN_WARNING  = 1
RTN_FAIL     = 2
RTN_CRITICAL = 3


# Module global vars
args = None
keylen = 0      # Padding vars used for pretty printing
hostlen = 0


def convert_time (timeX):
    """ Given time value term such as 90s, 15m, 20h, 3d, 2w return int seconds and units.
        <1h> returns (3600, "hours")
    If no units suffix, the value is interpreted as secs.
        <90> returns (90, "secs ")
    """
    if type(timeX) is int:
        return timeX, "secs "           # Case Int
    try:
        return int(timeX), "secs "      # Case Str without units
    except:
        pass
    try:                                # Case Str with units
        time_value = int(timeX[:-1])
        time_unit =  timeX[-1:].lower()
        if time_unit == "s":
            return time_value, "secs "
        if time_unit == "m":
            return time_value * 60, "mins "
        if time_unit == "h":
            return time_value * 60*60, "hours"
        if time_unit == "d":
            return time_value * 60*60*24, "days "
        if time_unit == "w":
            return time_value * 60*60*24*7, "weeks"
        raise ValueError(f"Illegal time units <{time_unit}>")
    except Exception as e:
        logging.error (f"ERROR:  Could not convert time value <{timeX}>\n  {e}")
        raise


USER_HOST_FORMAT = re.compile(r"[\w.-]+@([\w.-]+)$")
USER_HOST_PORT_FORMAT = re.compile(r"[\w.-]+@([\w.-]+):([\d]+)$")

def split_user_host_port(u_h_p):
    """ Handle variations in passed-in user_host_port (u_h_p)
        "local", "user@host", or "user@host:port"
    Return separate user_host, host, and port values
    EG:
        "xyz@host15:4455" returns:
            ["xyz@host15", "host15", "4455"]
        "xyz@host15" returns:
            ["xyz@host15", "host15", "22"]  (the default port is "22" for ssh)
        "local" returns
            ["local", "local", ""]
    """
    user_host_noport = u_h_p.split(":")[0]
    host = u_h_p
    port = ""
    if host != "local":
        port = "22"
        out = USER_HOST_FORMAT.match(u_h_p)
        if out:
            host = out.group(1)
        else:
            out = USER_HOST_PORT_FORMAT.match(u_h_p)
            if out:
                host = out.group(1)
                port = out.group(2)
            else:
                _msg = f"Expecting <user@host> or <user@host:port> format, but found <{u_h_p}>."
                logging.error(f"ERROR:  {_msg}")
                raise ValueError (_msg)
    return user_host_noport, host, port


def cmd_check(cmd, user_host_port, return_type=None, check_line_text=None, expected_text=None, not_text=None):
    """  Runs the cmd and operates on the response based on return_type.
    return_types:
        check_string        Returns True if expected_text occurs in response of the cmd, plus the full full subprocess run structure
            check_line_text     If provided, only the first line containing this text is checked
            expected_text       Text that must be found
            not_text            Text that must NOT be found
        cmdrun              Returns True if the cmd return code was 0, plus the full full subprocess run structure
    """

    if user_host_port != "local":
        u_h, _, port = split_user_host_port(user_host_port)
        cmd = ["ssh", u_h, "-p" + port, "-o", "ConnectTimeout=1", "-T"] + cmd

    for nTry in range (getcfg('nRetries')):
        try:
            logging.debug(f"cmd_check subprocess command try {nTry+1}: <{cmd}>")
            # runtry = subprocess.run(cmd, capture_output=True, text=True)  # Py 3.7+
            runtry = subprocess.run(cmd, stdout=subprocess.PIPE, stderr=subprocess.PIPE, universal_newlines=True)   #Py3.6 requires old-style params
        except Exception as e:
            logging.error(f"ERROR:  subprocess.run of cmd <{cmd}> failed.\n  {e}")
            return False, None

        if return_type == "check_string":
            if check_line_text is None:
                text_to_check = runtry.stdout
            else:
                text_to_check = ""
                for line in runtry.stdout.split("\n"):
                    if check_line_text in line:
                        text_to_check = line
                        break
                
            if expected_text in text_to_check:
                if not_text is not None:
                    if not_text not in text_to_check:
                        return True, runtry
                else:
                    return True, runtry
            else:
                if nTry == getcfg('nRetries')-1:
                    return False, runtry

        elif return_type == "cmdrun":
            if runtry.returncode == 0:
                return True, runtry
            elif nTry == getcfg('nRetries')-1:
                return False, runtry

        else:
            _msg = f"Invalid return_type <{return_type}> passed to cmd_check"
            logging.error (f"ERROR:  {_msg}")
            raise ValueError (_msg)
        time.sleep (convert_time(getcfg('RetryInterval'))[0])


have_access_dict = {}

def have_access(user_host_port):
    """If user_host_port != "local" then check ping and ssh access to the host.
    Remember passing result and avoid redoing the ping and ssh tests on later checks.
    If ping/ssh tests fail they will be retried if referenced in a later monitored item.
    The dictionary is cleared on each RecheckInterval in the lanmonitor main code.
    ping or ssh access fails return a tuple of (False, error_message_string).
    (True, "") is returned if local or if ping and ssh tests pass for non-local host.

    NOTE that this optimization can result in confusion:
        if access once passed then fails the plugin will be called and may flag an error
        while the actual problem is no access.
    """
    user_host_port = user_host_port.lower()
    if user_host_port == "local":
        return (True, "")
    if user_host_port in have_access_dict:
        return (True, "")

    _, host, _ = split_user_host_port(user_host_port)
    if host != "local":     # Check for remote access if non-local
        pingrslt = cmd_check(["ping", "-c", "1", "-W", "1", host], user_host_port="local", return_type="cmdrun")
        if not pingrslt[0]:
            return (False, "HOST CANNOT BE REACHED")
        else:
            sshrslt = cmd_check(["echo", "hello"], user_host_port=user_host_port, return_type="cmdrun")
            if "Connection refused" in sshrslt[1].stderr:
                return (False, f"SSH CONNECTION REFUSED <{user_host_port}>")
            # if "The authenticity of host" in sshrslt[1].stdout:       # Never happens thru the script, only interactive
            #     return (False, f"SSH CONNECTION NOT APPROVED.  ssh to <{user_host_port}> manually first.")

    have_access_dict[user_host_port] = True
    return (True, "")


def next_summary_timestring():
    """Calculate timestring of next summary.
    SummaryDays			0 1 2 3 4 5 6	# Days of week.  0 = Sunday
    SummaryTime			9:45		    # 24 hour clock
    NOTE:  May be off by 1 hour over a DTS changes.
    """
    try:
        target_hour   = int(getcfg("SummaryTime","").split(":")[0])
        target_minute = int(getcfg("SummaryTime","").split(":")[1])
        today_day_num = datetime.datetime.today().isoweekday()    # Sunday = 0, Saturday = 6
        now = datetime.datetime.now().replace(second=0, microsecond=0)
        next_summary = now + datetime.timedelta(days=30)
        for daynum in getcfg("SummaryDays").split():
            offset_num_days = int(daynum) - today_day_num
            if offset_num_days < 0:
                offset_num_days += 7
            plus_day =  now + datetime.timedelta(days=offset_num_days)
            with_time = plus_day.replace(hour=target_hour, minute=target_minute)
            if with_time < datetime.datetime.now():
                with_time = with_time + datetime.timedelta(days=7)
            if with_time < next_summary:
                next_summary = with_time
        logging.debug(f"Next summary:  {next_summary}")
        return next_summary
    except Exception as e:
        _msg = f"SummaryDays <{getcfg('SummaryDays','')}> or SummaryTime <{getcfg('SummaryTime','')}> settings could not be parsed\n  {e}"
        logging.error(f"ERROR:  {_msg}")
        raise ValueError (_msg) from None
        