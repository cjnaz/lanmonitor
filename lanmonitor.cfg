# LAN Monitor config file
#  220420  Updated to V1.4 (LogLevel and LogFile)


# Core tool params
nRetries            3               # Num times to try to read an item before erroring
RetryInterval       1s              # Wait time between failed item reads
StartupDelay        30s             # Wait time when starting in service mode to allow everything to come up fully at system boot before checking items
RecheckInterval     15m             # Wait time between main loops
Gateway             192.168.1.1     # IP or Hostname used for checking for network access as a gate for checks on non-local hosts
LogLevel            30              # Logging module levels: 10:DEBUG, 20:INFO, 30:WARNING (default), 40:ERROR, 50:CRITICAL
LogFile             <path/to>/logfile.txt    # absolute or relative (to script directory) path


# Whitespace separated list of notification handlers
Notif_handlers      stock_notif

# Stock notifications handler params
CriticalReNotificationInterval 	2h      # Wait time for critical event notification message resend
SummaryDays             0 1 2 3 4 5 6   # Days of week.  0 = Sunday, 6 = Saturday
SummaryTime             09:00           # 24 hour clock
LogSummary              True            # True: Force logging of summary


# Monitor types and items
MonType_SELinux		selinux_plugin
SELinux_local           local           enforcing       # or permissive

MonType_Host		pinghost_plugin
Host_Printer_Server     local           192.168.1.44

MonType_Service	    service_plugin
Service_lanmonitor      local           lanmonitor 
Service_routermonitor   local           routermonitor 
Service_wanstatus       local           wanstatus
Service_httpd           local           httpd 
Service_firewalld       local           CRITICAL    firewalld 
Service_smb	            local           CRITICAL    smb 
Service_sshd            local           sshd
Service_mariadb         local           mariadb
Service_RPi1_cups       me@RPi1.lan     CRITICAL    cups

MonType_Page        webpage_plugin
Page_WeeWX              local           http://localhost/weewx/         Current Conditions
Page_xBrowserSync       me@RPi2.mylan   https://www.xbrowsersync.org/   Browser syncing as it should be: secure, anonymous and free! # Check it out!

MonType_Process     process_plugin
Process_x11vnc          local           CRITICAL    /usr/bin/x11vnc

MonType_Activity    fsactivity_plugin
Activity_MyServer_backups local         8d     /mnt/share/MyServerBackups
Activity_RPi2_log.csv   rpi2.mylan      5m     /mnt/RAMDRIVE/log.csv

MonType_YumUpdate  yum_update_history_plugin
YumUpdate_myserver      local           CRITICAL  15d  update --skip-broken


# Email and Notification params
EmailFrom	        your.email@gmail.com
NotifList           4805551234@vzwpix.com   # For notifications - One or more addresses, whitespace separated.  Use your carrier's email-to-text bridge address.  Regular email addresses may be used.
EmailTo             your.email@gmail.com    # For perodic summaries - One or more addresses, whitespace separated.
Import              ~/creds_SMTP            # Provides EmailServer, EmailServerPort, EmailUser and EmailPass
#EmailVerbose	    True                    # True: enable the emailer debug level
#DontEmail          True                    # True: Emails (including notifications) will NOT be sent
#DontNotif          True                    # True: Notifications will NOT be sent
