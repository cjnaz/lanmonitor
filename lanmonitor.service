[Unit]
Description=Monitor processes and devices on the LAN
After=network-online.target

[Service]
ExecStart=/<path to>/lanmonitor

[Install]
WantedBy=multi-user.target
