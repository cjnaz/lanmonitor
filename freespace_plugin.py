#!/usr/bin/env python3
"""LAN Monitor plugin - freespace_plugin

Free space at the specified path is checked.  The `friendly_name` is user defined (not the real path).
Expected free space may be an absolute value or a percentage.

      MonType_Free		freespace_plugin
      Free_<friendly_name>  <local or user@host>  [CRITICAL]  <check_interval>  <free_amount>  <path>
      Free_RPi1    pi@RPi3    CRITICAL   5m   20%         /home/pi
      Free_share   local                 1d   1000000000  /mnt/share

If the path does not exist, setup() will pass but eval_status() will return RTN_WARNING and retry on each
check_interval.  This allows for intermittently missing paths.
"""

__version__ = "V2.0 221130"

#==========================================================
#
#  Chris Nelson, 2021-2022
#
# V2.0 221130  Initial
#
# Changes pending
#   
#==========================================================

import datetime
import re
import globvars
from lanmonfuncs import RTN_PASS, RTN_WARNING, RTN_FAIL, RTN_CRITICAL, cmd_check
from funcs3 import logging

# Configs / Constants
SPACE_RE = re.compile(r"\d+ +\d+ +(\d+ +) +(\d+)%")
# df output example:
#   Filesystem     1K-blocks    Used Available Use% Mounted on
#   /dev/root       15022928 2556068  11824848  18% /
# SPACE_RE picks up group(1)='11824848', group(2)='18'


class monitor:

    def __init__ (self):
        pass

    def setup (self, item):
        """ Set up instance vars and check item values.
        Passed in item dictionary keys:
            key             Full 'itemtype_tag' key value from config file line
            tag             'tag' portion only from 'itemtype_tag' from config file line
            user_host_port  'local' or 'user@hostname[:port]' from config file line
            host            'local' or 'hostname' from config file line
            critical        True if 'CRITICAL' is in the config file line
            check_interval  Time in seconds between rechecks
            rest_of_line    Remainder of line (plugin specific formatting)
        Returns True if all good, else False
        """

        logging.debug (f"{__name__}.setup()  called for  {item['key']}:\n  {item}")

        self.key            = item["key"]                           # vvvv These items don't need to be modified
        self.key_padded     = self.key.ljust(globvars.keylen)
        self.tag            = item["tag"]
        self.user_host_port = item["user_host_port"]
        self.host           = item["host"]
        self.host_padded    = self.host.ljust(globvars.hostlen)
        if item["critical"]:
            self.failtype = RTN_CRITICAL
            self.failtext = "CRITICAL"
        else:
            self.failtype = RTN_FAIL
            self.failtext = "FAIL"
        self.next_run       = datetime.datetime.now()
        self.check_interval = item['check_interval']                # ^^^^ These items don't need to be modified

        percent_form = re.search("^([\d]+)%\s+(.+)", item["rest_of_line"])
        if percent_form:
            self.minfree   = int(percent_form.group(1))
            self.free_type = 'percent'
            self.path = percent_form.group(2)
            return RTN_PASS

        else:
            absolute_form = re.search("^([\d]+)\s+(.+)", item["rest_of_line"])
            if absolute_form:
                self.minfree   = int(absolute_form.group(1))
                self.free_type = 'absolute'
                self.path = absolute_form.group(2)
                return RTN_PASS

        logging.error (f"  ERROR:  <{self.key}> COULD NOT PARSE SETTINGS <{item['rest_of_line']}>")
        return RTN_FAIL


    def eval_status (self):
        """ Check status of this item.
        Returns dictionary with these keys:
            rslt            Integer status:  RTN_PASS, RTN_WARNING, RTN_FAIL, RTN_CRITICAL
            notif_key       Unique handle for tracking active notifications in the notification handler 
            message         String with status and context details
        """

        logging.debug (f"{__name__}.eval_status()  called for  {self.key}")

        cmd = ["df", self.path]
        df_rslt = cmd_check(cmd, user_host_port=self.user_host_port, return_type="cmdrun")
        # logging.debug (f"cmd_check response:  {df_rslt}")
        if not df_rslt[0]:
            return {"rslt":RTN_WARNING, "notif_key":self.key, "message":f"  WARNING: {self.key} - {self.host} - COULD NOT GET df OF PATH <{self.path}>"}

        out = SPACE_RE.search(df_rslt[1].stdout)
        if out:
            if self.free_type == 'absolute':
                abs_free = int(out.group(1))
                if abs_free > self.minfree:
                    return {"rslt":RTN_PASS, "notif_key":self.key, "message":f"{self.key_padded}  OK - {self.host_padded} - free {abs_free}  (minfree {self.minfree})  {self.path}"}
                else:
                    return {"rslt":self.failtype, "notif_key":self.key, "message":f"  {self.failtext}: {self.key} - {self.host} - free {abs_free}  (minfree {self.minfree})  {self.path}"}
            else:   # percent case
                percent_free = 100 - int(out.group(2))
                if percent_free > self.minfree:
                    return {"rslt":RTN_PASS, "notif_key":self.key, "message":f"{self.key_padded}  OK - {self.host_padded} - free {percent_free}%  (minfree {self.minfree}%)  {self.path}"}
                else:
                    return {"rslt":self.failtype, "notif_key":self.key, "message":f"  {self.failtext}: {self.key} - {self.host} - free {percent_free}%  (minfree {self.minfree}%)  {self.path}"}
        else:
            logging.debug (f"df of <{self.path}> not parsable - returned\n  {df_rslt[1].stdout}")
            return {"rslt":RTN_WARNING, "notif_key":self.key, "message":f"  WARNING: {self.key} - {self.host} - df OF PATH <{self.path}> not parsable"}


if __name__ == '__main__':
    import argparse
    from funcs3 import loadconfig

    CONFIG_FILE = "lanmonitor.cfg"

    parser = argparse.ArgumentParser(description=__doc__ + __version__, formatter_class=argparse.RawTextHelpFormatter)
    parser.add_argument('--config-file', default=CONFIG_FILE,
                        help=f"Path to config file (default <{CONFIG_FILE}>).")
    parser.add_argument('-V', '--version', action='version', version='%(prog)s ' + __version__,
                        help="Return version number and exit.")

    globvars.args = parser.parse_args()
    loadconfig(cfgfile=globvars.args.config_file)
    logging.getLogger().setLevel(logging.DEBUG)


    def dotest (test):
        print ()
        inst = monitor()
        setup_rslt = inst.setup(test)
        print (f"  setup():  {setup_rslt}")
        if setup_rslt == RTN_PASS:
            print(f"  eval_status():  {inst.eval_status()}")

    dotest ({"key":"Free_Per_pass", "tag":"Per_pass", "host":"local", "user_host_port":"local", "critical":True, "check_interval":1, "rest_of_line":"30% /home"})

    dotest ({"key":"Free_Per_fail", "tag":"Per_fail", "host":"local", "user_host_port":"local", "critical":True, "check_interval":1, "rest_of_line":"100% /home"})

    dotest ({"key":"Free_Abs_pass", "tag":"Abs_pass", "host":"rpi3", "user_host_port":"pi@rpi3", "critical":True, "check_interval":1, "rest_of_line":"18000 /mnt/RAMDRIVE"})

    dotest ({"key":"Free_Abs_fail", "tag":"Abs_fail", "host":"rpi3", "user_host_port":"pi@rpi3", "critical":True, "check_interval":1, "rest_of_line":"30000 /mnt/RAMDRIVE"})

    dotest ({"key":"Free_nosuchpath", "tag":"nosuchpath", "host":"local", "user_host_port":"local", "critical":True, "check_interval":1, "rest_of_line":"10% /mnt/nosuchpath"})

    dotest ({"key":"Free_pathWithSpaces", "tag":"pathWithSpaces", "host":"local", "user_host_port":"local", "critical":True, "check_interval":1, "rest_of_line":"10% /mnt/share/tmp/Ripped videos"})
    
    dotest ({"key":"Free_badlimit", "tag":"badlimit", "host":"local", "user_host_port":"local", "critical":True, "check_interval":1, "rest_of_line":"10x% /home"})
    