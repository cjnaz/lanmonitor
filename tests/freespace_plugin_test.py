#!/usr/bin/env python3
"""LAN Monitor plugin - freespace_plugin_test
"""

#==========================================================
#
#  Chris Nelson, Copyright 2021-2023
#
# 3.1 230320 - Added ssh access warning cases
# 3.0 230301 - Packaged
#
#==========================================================

from cjnfuncs.cjnfuncs import set_toolname, setuplogging, logging, cfg
from lanmonitor.lanmonfuncs import RTN_PASS
try:
    from freespace_plugin import monitor
except:
    from lanmonitor.freespace_plugin import monitor

tool = set_toolname("tool")
cfg["nRetries"]         = 1
cfg["RetryInterval"]    = "0s"
cfg["ConsoleLogFormat"] = "{module:>35}.{funcName:20} - {levelname:>8}:  {message}"
setuplogging()
logging.getLogger().setLevel(logging.DEBUG)

test_num = 0
def dotest (test):
    global test_num
    test_num += 1
    print(f"\nTest {test_num} {'-'*50}")
    inst = monitor()
    setup_rslt = inst.setup(test)
    logging.debug (f"{test['key']} - setup() returned:  {setup_rslt}")
    if setup_rslt == RTN_PASS:
        logging.debug (f"{test['key']} - eval_status() returned:  {inst.eval_status()}")

dotest ({"key":"Free_Per_pass", "tag":"Per_pass", "host":"local", "user_host_port":"local", "critical":True, "check_interval":1, "rest_of_line":"30% /home"})

dotest ({"key":"Free_Per_fail", "tag":"Per_fail", "host":"local", "user_host_port":"local", "critical":True, "check_interval":1, "rest_of_line":"100% /home"})

dotest ({"key":"Free_Abs_pass", "tag":"Abs_pass", "host":"rpi3", "user_host_port":"pi@rpi3", "critical":True, "check_interval":1, "rest_of_line":"18000 /mnt/RAMDRIVE"})

dotest ({"key":"Free_Abs_fail", "tag":"Abs_fail", "host":"rpi3", "user_host_port":"pi@rpi3", "critical":True, "check_interval":1, "rest_of_line":"30000 /mnt/RAMDRIVE"})

dotest ({"key":"Free_nosuchpath", "tag":"nosuchpath", "host":"local", "user_host_port":"local", "critical":True, "check_interval":1, "rest_of_line":"10% /mnt/nosuchpath"})

dotest ({"key":"Free_pathWithSpaces", "tag":"pathWithSpaces", "host":"local", "user_host_port":"local", "critical":True, "check_interval":1, "rest_of_line":"10% /mnt/share/tmp/Ripped videos"})

dotest ({"key":"Free_badlimit", "tag":"badlimit", "host":"local", "user_host_port":"local", "critical":True, "check_interval":1, "rest_of_line":"10x% /home"})

dotest ({"key":"Free_Unknown", "tag":"Unknown", "host":"nosuchhost", "user_host_port":"pi@nosuchhost", "critical":True, "check_interval":1, "rest_of_line":"30% /home"})

dotest ({"key":"Free_Unavailable", "tag":"Unavailable", "host":"shopcam", "user_host_port":"me@shopcam", "critical":True, "check_interval":1, "rest_of_line":"30% /home"})
