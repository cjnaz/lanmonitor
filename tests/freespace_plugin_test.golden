
Test 1 - Freespace as a percentage - OK --------------------------------------------------
                   freespace_plugin.setup                -    DEBUG:  Free_Per_pass - lanmonitor.freespace_plugin.setup() called:
  {'key': 'Free_Per_pass', 'tag': 'Per_pass', 'host': 'local', 'user_host_port': 'local', 'critical': True, 'cmd_timeout': 2, 'check_interval': 1, 'rest_of_line': '30% /home   '}
              freespace_plugin_test.dotest               -    DEBUG:  Free_Per_pass - setup() returned:  0
                   freespace_plugin.eval_status          -    DEBUG:  Free_Per_pass - lanmonitor.freespace_plugin.eval_status() called
                        lanmonfuncs.cmd_check            -    DEBUG:  cmd try 1 <['df', '/home']>
              freespace_plugin_test.dotest               -    DEBUG:  Free_Per_pass - eval_status() returned:  {'rslt': 0, 'notif_key': 'Free_Per_pass', 'message': 'Free_Per_pass  OK - local - free 91%  (minfree 30%)  /home'}

Test 2 - Freespace as a percentage - CRITICAL --------------------------------------------------
                   freespace_plugin.setup                -    DEBUG:  Free_Per_fail - lanmonitor.freespace_plugin.setup() called:
  {'key': 'Free_Per_fail', 'tag': 'Per_fail', 'host': 'local', 'user_host_port': 'local', 'critical': True, 'cmd_timeout': 2, 'check_interval': 1, 'rest_of_line': '100% /home'}
              freespace_plugin_test.dotest               -    DEBUG:  Free_Per_fail - setup() returned:  0
                   freespace_plugin.eval_status          -    DEBUG:  Free_Per_fail - lanmonitor.freespace_plugin.eval_status() called
                        lanmonfuncs.cmd_check            -    DEBUG:  cmd try 1 <['df', '/home']>
              freespace_plugin_test.dotest               -    DEBUG:  Free_Per_fail - eval_status() returned:  {'rslt': 3, 'notif_key': 'Free_Per_fail', 'message': '  CRITICAL: Free_Per_fail - local - free 91%  (minfree 100%)  /home'}

Test 3 - Freespace as a absolute min remote - OK --------------------------------------------------
                   freespace_plugin.setup                -    DEBUG:  Free_Abs_pass - lanmonitor.freespace_plugin.setup() called:
  {'key': 'Free_Abs_pass', 'tag': 'Abs_pass', 'host': 'testhost2', 'user_host_port': 'me@testhost2', 'critical': True, 'cmd_timeout': 2, 'check_interval': 1, 'rest_of_line': '1000 /mnt/RAMDRIVE   '}
              freespace_plugin_test.dotest               -    DEBUG:  Free_Abs_pass - setup() returned:  0
                   freespace_plugin.eval_status          -    DEBUG:  Free_Abs_pass - lanmonitor.freespace_plugin.eval_status() called
                        lanmonfuncs.cmd_check            -    DEBUG:  cmd try 1 <['ssh', 'me@testhost2', '-p22', '-T', 'df', '/mnt/RAMDRIVE']>
              freespace_plugin_test.dotest               -    DEBUG:  Free_Abs_pass - eval_status() returned:  {'rslt': 0, 'notif_key': 'Free_Abs_pass', 'message': 'Free_Abs_pass  OK - testhost2 - free 20476  (minfree 1000)  /mnt/RAMDRIVE'}

Test 4 - Freespace as a absolute min remote - CRITICAL --------------------------------------------------
                   freespace_plugin.setup                -    DEBUG:  Free_Abs_fail - lanmonitor.freespace_plugin.setup() called:
  {'key': 'Free_Abs_fail', 'tag': 'Abs_fail', 'host': 'testhost2', 'user_host_port': 'me@testhost2', 'critical': True, 'cmd_timeout': 2, 'check_interval': 1, 'rest_of_line': '300000 /mnt/RAMDRIVE'}
              freespace_plugin_test.dotest               -    DEBUG:  Free_Abs_fail - setup() returned:  0
                   freespace_plugin.eval_status          -    DEBUG:  Free_Abs_fail - lanmonitor.freespace_plugin.eval_status() called
                        lanmonfuncs.cmd_check            -    DEBUG:  cmd try 1 <['ssh', 'me@testhost2', '-p22', '-T', 'df', '/mnt/RAMDRIVE']>
              freespace_plugin_test.dotest               -    DEBUG:  Free_Abs_fail - eval_status() returned:  {'rslt': 3, 'notif_key': 'Free_Abs_fail', 'message': '  CRITICAL: Free_Abs_fail - testhost2 - free 20476  (minfree 300000)  /mnt/RAMDRIVE'}

Test 5 - Known host, No such path - WARNING --------------------------------------------------
                   freespace_plugin.setup                -    DEBUG:  Free_nosuchpath - lanmonitor.freespace_plugin.setup() called:
  {'key': 'Free_nosuchpath', 'tag': 'nosuchpath', 'host': 'local', 'user_host_port': 'local', 'critical': True, 'cmd_timeout': 2, 'check_interval': 1, 'rest_of_line': '10% /mnt/nosuchpath'}
              freespace_plugin_test.dotest               -    DEBUG:  Free_nosuchpath - setup() returned:  0
                   freespace_plugin.eval_status          -    DEBUG:  Free_nosuchpath - lanmonitor.freespace_plugin.eval_status() called
                        lanmonfuncs.cmd_check            -    DEBUG:  cmd try 1 <['df', '/mnt/nosuchpath']>
                        lanmonfuncs.cmd_check            -    DEBUG:  cmd try 1 failed (returncode=1) <['df', '/mnt/nosuchpath']>:  df: /mnt/nosuchpath: No such file or directory
              freespace_plugin_test.dotest               -    DEBUG:  Free_nosuchpath - eval_status() returned:  {'rslt': 1, 'notif_key': 'Free_nosuchpath', 'message': '  WARNING: Free_nosuchpath - local - COULD NOT GET df OF PATH </mnt/nosuchpath>'}

Test 6 - Space in path - OK --------------------------------------------------
                   freespace_plugin.setup                -    DEBUG:  Free_pathWithSpaces - lanmonitor.freespace_plugin.setup() called:
  {'key': 'Free_pathWithSpaces', 'tag': 'pathWithSpaces', 'host': 'local', 'user_host_port': 'local', 'critical': True, 'cmd_timeout': 2, 'check_interval': 1, 'rest_of_line': '10% /mnt/share/tmp/Ripped videos'}
              freespace_plugin_test.dotest               -    DEBUG:  Free_pathWithSpaces - setup() returned:  0
                   freespace_plugin.eval_status          -    DEBUG:  Free_pathWithSpaces - lanmonitor.freespace_plugin.eval_status() called
                        lanmonfuncs.cmd_check            -    DEBUG:  cmd try 1 <['df', '/mnt/share/tmp/Ripped videos']>
              freespace_plugin_test.dotest               -    DEBUG:  Free_pathWithSpaces - eval_status() returned:  {'rslt': 0, 'notif_key': 'Free_pathWithSpaces', 'message': 'Free_pathWithSpaces  OK - local - free 31%  (minfree 10%)  /mnt/share/tmp/Ripped videos'}

Test 7 - Bad limit value - setup ERROR --------------------------------------------------
                   freespace_plugin.setup                -    DEBUG:  Free_badlimit - lanmonitor.freespace_plugin.setup() called:
  {'key': 'Free_badlimit', 'tag': 'badlimit', 'host': 'local', 'user_host_port': 'local', 'critical': True, 'cmd_timeout': 2, 'check_interval': 1, 'rest_of_line': '10x% /home'}
                   freespace_plugin.setup                -    ERROR:    ERROR:  <Free_badlimit> COULD NOT PARSE SETTINGS <10x% /home>
              freespace_plugin_test.dotest               -    DEBUG:  Free_badlimit - setup() returned:  2

Test 8 - No such host - WARNING --------------------------------------------------
                   freespace_plugin.setup                -    DEBUG:  Free_Unknown - lanmonitor.freespace_plugin.setup() called:
  {'key': 'Free_Unknown', 'tag': 'Unknown', 'host': 'nosuchhost', 'user_host_port': 'me@nosuchhost', 'critical': True, 'cmd_timeout': 2, 'check_interval': 1, 'rest_of_line': '30% /home'}
              freespace_plugin_test.dotest               -    DEBUG:  Free_Unknown - setup() returned:  0
                   freespace_plugin.eval_status          -    DEBUG:  Free_Unknown - lanmonitor.freespace_plugin.eval_status() called
                        lanmonfuncs.cmd_check            -    DEBUG:  cmd try 1 <['ssh', 'me@nosuchhost', '-p22', '-T', 'df', '/home']>
                        lanmonfuncs.cmd_check            -    DEBUG:  cmd try 1 failed (returncode=255) <['ssh', 'me@nosuchhost', '-p22', '-T', 'df', '/home']>:  ssh: Could not resolve hostname nosuchhost: Name or service not known
                        lanmonfuncs.cmd_check            -    DEBUG:  cmd failed on remote system with <ssh: Could not resolve hostname nosuchhost: Name or service not known> - attempting simplessh connection to <me@nosuchhost>
                        lanmonfuncs.cmd_check            -    DEBUG:  simplessh try 1 <['ssh', 'me@nosuchhost', '-p22', '-T', 'echo', 'hello']>
                        lanmonfuncs.cmd_check            -    DEBUG:  simplessh try 1 failed (returncode=255) <['ssh', 'me@nosuchhost', '-p22', '-T', 'echo', 'hello']>:  ssh: Could not resolve hostname nosuchhost: Name or service not known
                        lanmonfuncs.cmd_check            -    DEBUG:  simplessh failed with <ssh: Could not resolve hostname nosuchhost: Name or service not known>
              freespace_plugin_test.dotest               -    DEBUG:  Free_Unknown - eval_status() returned:  {'rslt': 1, 'notif_key': 'Free_Unknown', 'message': '  WARNING: Free_Unknown - nosuchhost - ssh: Could not resolve hostname nosuchhost: Name or service not known'}

Test 9 - Known host, unavailable - WARNING --------------------------------------------------
                   freespace_plugin.setup                -    DEBUG:  Free_Unavailable - lanmonitor.freespace_plugin.setup() called:
  {'key': 'Free_Unavailable', 'tag': 'Unavailable', 'host': 'testhostX', 'user_host_port': 'me@testhostX', 'critical': True, 'cmd_timeout': 2, 'check_interval': 1, 'rest_of_line': '30% /home'}
              freespace_plugin_test.dotest               -    DEBUG:  Free_Unavailable - setup() returned:  0
                   freespace_plugin.eval_status          -    DEBUG:  Free_Unavailable - lanmonitor.freespace_plugin.eval_status() called
                        lanmonfuncs.cmd_check            -    DEBUG:  cmd try 1 <['ssh', 'me@testhostX', '-p22', '-T', 'df', '/home']>
                        lanmonfuncs.cmd_check            -    DEBUG:  cmd try 1 timeout:  Command '['ssh', 'me@testhostX', '-p22', '-T', 'df', '/home']' timed out after 2 seconds
                        lanmonfuncs.cmd_check            -    DEBUG:  cmd failed on remote system with <Command '['ssh', 'me@testhostX', '-p22', '-T', 'df', '/home']' timed out after 2 seconds> - attempting simplessh connection to <me@testhostX>
                        lanmonfuncs.cmd_check            -    DEBUG:  simplessh try 1 <['ssh', 'me@testhostX', '-p22', '-T', 'echo', 'hello']>
                        lanmonfuncs.cmd_check            -    DEBUG:  simplessh try 1 failed (returncode=255) <['ssh', 'me@testhostX', '-p22', '-T', 'echo', 'hello']>:  ssh: connect to host testhostx port 22: No route to host
                        lanmonfuncs.cmd_check            -    DEBUG:  simplessh failed with <ssh: connect to host testhostx port 22: No route to host>
              freespace_plugin_test.dotest               -    DEBUG:  Free_Unavailable - eval_status() returned:  {'rslt': 1, 'notif_key': 'Free_Unavailable', 'message': '  WARNING: Free_Unavailable - testhostX - ssh: connect to host testhostx port 22: No route to host'}
